### ISSUE TITLE
New tests: passwords and english words

### ISSUE THREAD

**Issue by avaneev** [rurban/smhasher#137]
Hello! It would be useful to collision-test hash function against a common application: password storage.
E.g. on generated list of 300000 random 6-16-character passwords from the set ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789.,!?:;-+=()<>/\|"'@#$%&*_^`{}[]
This is what password generators usually produce. The mix of different-length passwords is an important part of the test.

Second test is a collision test on a vocabulary of English words, at least 200000 required.

**Comment by rurban** [comment]
Agreed, that sounds useful. Similar tests are under Text, but differently. Maybe Words

**Comment by avaneev** [comment]
For Words test it would be useful to randomize word length as well, from 4 to 12 chars.

**Comment by wangyi-fudan** [comment]
hi, I see both wyhash, xxh3, t1ha fails some word test. Is there a bug or you are sure of the results? if so, we should improve the hash function.

**Comment by rurban** [comment]
王一 WangYi <notifications@github.com> schrieb am Mi., 19. Aug. 2020, 02:17:

> hi, I see both wyhash, xxh3, t1ha fails some word test. Is there a bug or
> you are sure of the results? if so, we should improve the hash function.
>
No, there's certainly a bug in the collision counting in the new Words
test. Working on it. I've crosschecked with sort | uniq -c and it's wrong.
The hash functions are fine.
There are also more minor bugs in the tests lurking.

I also improved the collision estimation calc, which detoriated with large
numbers. Now it looks much better. The good hashes are now all perfect as
they should be.

**Comment by rurban** [comment]
Found my Words bug. It was a thinko. 200000 randomly created words certainly do have duplicates, so they report collisions. I have eliminate all duplicates from the initial words of course.

**Comment by avaneev** [comment]
You've removed a mix of different-length passwords from the test, this is not as reliable.

**Comment by rurban** [comment]
But needs much less memory. Using a list of random len strings is not much different than testing most len's seperately.
But it would be faster to test much less, of course, with a single long list of random-len strings. Thinking about it.

More important is to find out why some quality hashes fail this test now: 
Probably still wrong estimation calc with a very low number of colls (~4 colls on 32bit), loosing a lot of accuracy in the pow().

**Comment by avaneev** [comment]
I think different-length password test is important: it checks for hash distribution irrespective of message length.

**Comment by avaneev** [comment]
The reason these otherwise "good" hashes may fail on newer Words tests is that alphabet is restricted. Probably these hashes are tuned to hash full 8-bit data.

**Comment by avaneev** [comment]
You should probably allow 3x collisions for Words tests, on high-low bit ranges. For Combination tests I would also increase collision allowance depending on the hash length. For example, combination 0x800000000000000 test is too sparse for 128-bit hash to expect individual bit ranges will be collision-free. SHA256 also barely passes it, which is to be expected. You should probably find a formula which estimates entropy of input data, and adjust allowances according to it.

**Comment by rurban** [comment]
Certainly not. All good hashes must pass all statistical tests, for all bytes 0-255. Specializing in alnum might be a dirty hack for poor hashes only.

The problem is missing accuracy on the estimation calculation in corner cases, on low ests. The numbers differ too much, and errors propagate. We could do proper counting of the probabilities (slow, needs to be cached) or improve our estimator for the best cases. Still looking.

**Comment by avaneev** [comment]
Collision testing is not very reliable on hash parts. It can go 3x the expected value. Currently you allow 2x. I suggest you to increase allowance to 3x. This will solve the problem.

**Comment by rurban** [comment]
Nope. 3x 1000 is way too high. For low estimates (with high errors) I now allow factor 4, not 2. This should work

**Comment by avaneev** [comment]
That's good.
